src/submission/schema-migration/migration-manager.ts

```diff
@@ -172,80 +182,108 @@ const getNextUncheckedDonorDocumentsBatch = async (migrationId: string, limit: n
 };
 
 // TODO: enhance to return invalidation reasons
-const shouldInvalidate = (donor: DeepReadonly<Donor>, criteria: any) => {
-  let flag = false;
-  criteria.forEach((c: any) => {
-    // short circuit if already found invalidating criteria
-    if (flag) return;
-
-    if (c.type == 'CODELIST_UPDATED') {
-      const hasInvalidValue = donorHasInvalidValue(c.fieldPath, donor, c.noLongerValid);
-      if (hasInvalidValue) {
-        flag = true;
-        return;
-      }
-    }
+const revalidateDonorClinicalEntities = async (
+  donor: DeepReadonly<Donor>,
+  schema: SchemasDictionary,
+  newSchemaVersion: string,
+) => {
+  const donorSchemaErrors: any[] = [];
+  // analyze changes between the document last valid schema
+  const analysis = await manager
+    .instance()
+    .analyzeChanges(donor.schemaMetadata.lastValidSchemaVersion, newSchemaVersion);
 
-    if (c.type == 'CODELIST_ADDED') {
-      const hasInvalidValue = donorHasInvalidValue(c.fieldPath, donor, c.noLongerValid);
-      if (hasInvalidValue) {
-        flag = true;
-        return;
-      }
-    }
-  });
+  // check for breaking changes
+  const invalidatingFields: any = findInvalidatingChangesFields(analysis);
+
+  const schemaNamesWithBreakingChanges = _.uniqBy(
+    invalidatingFields.map((inf: any) => {
+      return inf.fieldPath.split('.')[0];
+    }),
+    (e: string) => e,
+  );
 
-  return flag;
+  for (const schemaName of schemaNamesWithBreakingChanges) {
```

Comment:
looks like we run the schema validation even if its just one breaking change, so wouldn't it be easier to just find only the `clinical_type` with breaking changes. do we need to find the specific fields here?? maybe in analysis we can keep track of schema names that are not valid?

Reply:
I checked the reason, we need to analyze by field first to see if it required or not in some cases check line 386 for example in this file

Reply:
fair enough, and looks like `breakingChangesEntitesCache` is keeping track of `clinical_types` between version changes so :+1: